### PR TITLE
Default Vault environment variables

### PR DESCRIPTION
**Pull request by leominov**
Built-in Vault environment variables.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Hi @leominov!  Thanks for the contribution.

When making such large changes, its usually a good idea to open an issue first, so we can discuss them.  I think this PR should probably be broken up into a bunch of smaller ones, but I'll try and comment on it as it.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,32 @@
...
+package/
+.idea/
+.vagrant/
 vault_exporter
+.uptodate
+.build
+dependencies-stamp
+.tarballs/
+coverage.out
+debug.test
```
Comment (tomwilkie) [inline .gitignore]: Most of this .gitignore seems unrelated to this project; which bits are necessary?

File: .promu.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.BuildDate={{date "20060102-15:04:05"}}
+tarball:
+  files:
+    - LICENSE
+crossbuild:
+  platforms:
+    - linux/amd64
+    - linux/386
+    - darwin/amd64
+    - darwin/386
```
Comment (tomwilkie) [inline .promu.yml]: Please put all the promu changes in a separate PR, and make it clear _why_ you want to use promu.  In general, I'm not a fan - I've had all sort of problems with it in the past, and I generally don't see why its necessary.

File: README.md [hunk 3]
```diff
@@ -52,15 +52,28 @@ Flags:
                           Address to listen on for web interface and telemetry.
       --web.telemetry-path="/metrics"  
                           Path under which to expose metrics.
-      --vault-tls-cacert=VAULT-TLS-CACERT  
-                          The path to a PEM-encoded CA cert file to use to verify the Vault server SSL certificate.
-      --vault-tls-client-cert=VAULT-TLS-CLIENT-CERT  
-                          The path to the certificate for Vault communication.
-      --vault-tls-client-key=VAULT-TLS-CLIENT-KEY  
-                          The path to the private key for Vault communication.
-      --insecure-ssl      Set SSL to ignore certificate validation.
```
Comment (tomwilkie) [inline README.md]: These flags were added in #2 and I'm not sure @Cryson would be very happy about you removing them without asking first...

File: README.md [hunk 4]
```diff
@@ -52,15 +52,28 @@ Flags:
...
+* `VAULT_CAPATH` – empty
+* `VAULT_CLIENT_CERT` – empty
+* `VAULT_CLIENT_KEY` – empty
+* `VAULT_CLIENT_TIMEOUT` – `0`
+* `VAULT_SKIP_VERIFY` – `0`
+* `VAULT_TLS_SERVER_NAME` – empty
+* `VAULT_WRAP_TTL` – not used (?)
+* `VAULT_MAX_RETRIES` – `0`
+* `VAULT_TOKEN` – empty
+* `VAULT_MFA` – not used (?)
```
Comment (tomwilkie) [inline README.md]: If we're going to list the env variables here, can we explain which ones should be set, why, and to what values?  And we can omit ones that at not used.

File: VERSION [hunk 5]
```diff
@@ -0,0 +1 @@
+0.1.1
```
Comment (tomwilkie) [inline VERSION]: Why have you put the version in a file?  Version is indicated by a git tag.  Duplicating it is unnecessary and will just lead to it being out of sync.

File: main.go [hunk 6]
```diff
@@ -13,25 +13,13 @@ import (
...
-		Default("/metrics").String()
-	vaultCACert = kingpin.Flag("vault-tls-cacert",
-		"The path to a PEM-encoded CA cert file to use to verify the Vault server SSL certificate.").String()
-	vaultClientCert = kingpin.Flag("vault-tls-client-cert",
-		"The path to the certificate for Vault communication.").String()
-	vaultClientKey = kingpin.Flag("vault-tls-client-key",
-		"The path to the private key for Vault communication.").String()
-	sslInsecure = kingpin.Flag("insecure-ssl",
-		"Set SSL to ignore certificate validation.").
-		Default("false").Bool()
```
Comment (tomwilkie) [inline main.go]: These flags were added in #2 and I'm not sure @Cryson would be very happy about you removing them without asking first...

File: main.go [hunk 7]
```diff
@@ -169,14 +138,14 @@ func main() {
...
-             <head><title>Vault Exporter</title></head>
-             <body>
-             <h1>Vault Exporter</h1>
-             <p><a href='` + *metricsPath + `'>Metrics</a></p>
-             <h2>Build</h2>
-             <pre>` + version.Info() + ` ` + version.BuildContext() + `</pre>
-             </body>
-             </html>`))
+			<head><title>Vault Exporter</title></head>
+			<body>
```
Comment (tomwilkie) [inline main.go]: Indentation looks broken here.

File: main.go (hunk 6 above)
Comment (leominov) [reply main.go]: Flags its duplicate of existing functionality:
* https://github.com/hashicorp/vault/blob/master/api/client.go#L123
* https://github.com/hashicorp/vault/blob/master/api/client.go#L198

File: VERSION (hunk 5 above)
Comment (leominov) [reply VERSION]: Ok, i'll try with Promu without this file.

File: README.md (hunk 4 above)
Comment (leominov) [reply README.md]: Ok

File: .gitignore (hunk 1 above)
Comment (leominov) [reply .gitignore]: ok, i'll clean

**Comment by leominov** [comment]
I'll think PR is pretty clean, @tomwilkie

File: main.go (hunk 6 above)
Comment (tomwilkie) [reply main.go]: I understand that - see https://github.com/grapeshot/vault_exporter/pull/2#issuecomment-373549504. Having multiple ways of doing the same thing isn't inherently bad though - for some people, managing flags might be easier than environment variables.

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,23 @@
+# Compiled Object files, Static and Dynamic libs (Shared Objects)
+*.o
+*.a
+*.so
+.DS_Store
+Thumbs.db
```
Comment (tomwilkie) [inline .gitignore]: Where does anything in this repository generate a `Thumbs.db`, of a `.o`, `.a`, `.so`?  Why are they here?

File: .gitignore [hunk 9]
```diff
@@ -1,2 +1,23 @@
...
+
+# Folders
+_obj
+_test
```
Comment (tomwilkie) [inline .gitignore]: Where does anything in this repository generate a `_obj` or `_test` folder?

File: .gitignore [hunk 10]
```diff
@@ -1,2 +1,23 @@
...
+
+# Architecture specific extensions/prefixes
+*.[568vq]
```
Comment (tomwilkie) [inline .gitignore]: Where does anything in this repository generate files with these extensions?

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,23 @@
...
+[568vq].out
+
+*.cgo1.go
+*.cgo2.c
+_cgo_defun.c
+_cgo_gotypes.go
+_cgo_export.*
```
Comment (tomwilkie) [inline .gitignore]: AFAICT, we don't use cgo in this repo.

File: .gitignore (hunk 8 above)
Comment (leominov) [reply .gitignore]: It's a default recommended Github .gitignore for golang-project.

File: .gitignore (hunk 9 above)
Comment (leominov) [reply .gitignore]: https://github.com/grapeshot/vault_exporter/pull/8#discussion_r182998801

**Comment by leominov** [comment]
@tomwilkie, I understand that this changes not so necessary at this time :) Thanks for your comment, next time I'll be more attentive.

I'll create another PR with EnvVar documentation.
